### DISCUSSION TITLE
Impossible to specify storage buffers as write-only?

### DISCUSSION THREAD

**Discussion by apessino** [gfx-rs/wgpu#2121] [General]
Hello!

I have a storage buffer that is used as an output in a render pass - WGSL allows me to specify its class as write-only, like so:

```
[[block]]
struct Outs {
    data: array<Out>;
};
[[group(1), binding(0)]]
var<storage, write> outs: Outs;
```

I can't find a way to specify the corresponding storage class on the native side - I can set the storage buffer to read only or read write, but not write only given the structure of `BindingType::Storage`

```
let layout1 = wgpu
            .device
            .create_bind_group_layout(&wgpu::BindGroupLayoutDescriptor {
                label: Some("surface_layout_group_1"),
                entries: &[wgpu::BindGroupLayoutEntry {
                    binding: 0,
                    visibility: wgpu::ShaderStages::FRAGMENT,
                    ty: wgpu::BindingType::Buffer {
                        ty: wgpu::BufferBindingType::Storage { read_only: false },
                        has_dynamic_offset: false,
                        min_binding_size: wgpu::BufferSize::new(SurfaceOutRe::re_size()),
                    },
                    count: None,
                }],
            });
```

This of course results in an error when I build the pipeline since the storage class is mismatched. For now I have set the buffer to `read_write` in the shader and this works fine, but I am not sure if I am missing how to specify a write-only class in native, or if indeed there is no way to do so. Since WGSL supports write-only I think it would make sense for wgpu to also have support.

Thanks! :)

A=

**Comment by kvark** [comment]
WGSL doesn't support write-only storage buffers, at least not yet. If you look at https://gpuweb.github.io/gpuweb/wgsl/#storage-class table, it lists the allowed access qualifiers for each class. Storage class can only be `read` or ` read_write`.

**Reply by apessino** [reply]
Indeed you are correct and I must have misread the spec - thanks for the clarification!

Still, it is interesting that `var<storage, write>` does not produce an error, and indeed it seems that as far as the shader is concerned the buffer is indeed write only. The error that gets generated at pipeline creation is:

```
error matching FRAGMENT shader requirements against the pipeline
    shader global ResourceBinding { group: 1, binding: 0 } is not available in the layout pipeline layout
    storage class Storage { access: LOAD | STORE } doesn't match the shader Storage { access: STORE }
```

It does appear that WGSL is good to go, but the WGPU side is not. 

Thanks again for the quick reply!

A=

**Reply by kvark** [reply]
Something should fail validation before it reaches the pipeline creation. We are just missing this check.
